### PR TITLE
Part 4: feature/binary input basic cluster

### PR DESCRIPTION
**Pull request by ghost**
#### Problem
The matter project has currently no support for the binary input (basic) cluster. This PR adds a support for that.

#### Change overview
* added emberAfBinaryInputBasicClusterServerInitCallback support (this will need an update of zap repo ->  relies on https://github.com/project-chip/zap/pull/154)
* added binary-input-basic-server.cpp impl.
* added binary input basic to chip-device-ctl (relies on https://github.com/project-chip/zap/pull/146)

#### Limitations
* used a custom zap tool because of https://github.com/project-chip/zap/issues/155 used version https://github.com/eve-cxrp/zap/tree/feature/sensor_clusters
* zap tool needs to be updated with fixes mentioned above
* depends on https://github.com/project-chip/connectedhomeip/pull/7189

#### Testing
Tested on custom accessory with contact sensor. Test setup was similar to TE2.

#### HW Platform
Nordic nRF52840

#### Notes
Related PRs
* https://github.com/project-chip/connectedhomeip/pull/7026
* https://github.com/project-chip/connectedhomeip/pull/7028
* https://github.com/project-chip/connectedhomeip/pull/7030

I have also created a issue regarding the name.
* https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/3268

### CODE REVIEW

File: src/app/zap_cluster_list.py [hunk 1]
```diff
@@ -9,13 +9,15 @@
 
 # List of directories in src/app/clusters to build for server clusters.
 SERVER_CLUSTERS: typing.Dict[str, typing.List[str]] = {
+    'AIR_PRESSURE_MEASUREMENT_CLUSTER': ['air-pressure-measurement-server'],
```
Comment (vivien-apple) [inline src/app/zap_cluster_list.py:12]: nit: Remove this line

File: src/app/zap_cluster_list.py [hunk 2]
```diff
@@ -44,6 +46,7 @@
     'OTA_CLIENT_CLUSTER': [],
     'POWER_CONFIG_CLUSTER': [],
     'PUMP_CONFIG_CONTROL_CLUSTER': [],
+    'RELATIVE_HUMIDITY_MEASUREMENT_CLUSTER': ['relative-humidity-measurement-server'],
```
Comment (vivien-apple) [inline src/app/zap_cluster_list.py:49]: nit: remove this line

File: src/app/zap_cluster_list.py [hunk 3]
```diff
@@ -61,13 +64,15 @@
 
 # List of directories in src/app/clusters to build for client clusters.
 CLIENT_CLUSTERS: typing.Dict[str, typing.List[str]] = {
+    'AIR_PRESSURE_MEASUREMENT_CLUSTER': [],
```
Comment (vivien-apple) [inline src/app/zap_cluster_list.py:67]: nit: Remove this line

File: src/app/zap_cluster_list.py [hunk 4]
```diff
@@ -96,6 +101,7 @@
     'OTA_CLIENT_CLUSTER': [],
     'POWER_CONFIG_CLUSTER': [],
     'PUMP_CONFIG_CONTROL_CLUSTER': [],
+    'RELATIVE_HUMIDITY_MEASUREMENT_CLUSTER': [],
```
Comment (vivien-apple) [inline src/app/zap_cluster_list.py:104]: nit: Remove this line

File: src/controller/data_model/controller-clusters.zap [hunk 5]
```diff
@@ -735,6 +735,103 @@
             }
           ]
         },
+        {
+          "name": "Binary Input (Basic)",
```
Comment (vivien-apple) [inline src/controller/data_model/controller-clusters.zap:739]: Can you remove the `Silabs` definitions from `src/app/zap-templates/zlc/data_model/Silabs/general.xml` and `src/app/zcl/zap-templates/data_model/Silabs/general-thread.xml` and add your own into `src/app/zap-templates/zcl/input-output-value.xml` ? It will be easier to compare the definition to the spec.

File: src/controller/data_model/gen/CHIPClientCallbacks.h [hunk 6]
```diff
@@ -158,6 +156,3 @@ typedef void (*TvChannelTvChannelListListAttributeCallback)(void * context, uint
 typedef void (*TargetNavigatorTargetNavigatorListListAttributeCallback)(void * context, uint16_t count,
                                                                         _NavigateTargetTargetInfo * entries);
 typedef void (*TestClusterListInt8uListAttributeCallback)(void * context, uint16_t count, uint8_t * entries);
-typedef void (*TestClusterListOctetStringListAttributeCallback)(void * context, uint16_t count, chip::ByteSpan * entries);
-typedef void (*TestClusterListStructOctetStringListAttributeCallback)(void * context, uint16_t count,
-                                                                      _TestListStructOctet * entries);
```
Comment (vivien-apple) [inline src/controller/data_model/gen/CHIPClientCallbacks.h:163]: Seems like unrelated changes.

Comment (eve-cxrp) [reply src/controller/data_model/gen/CHIPClientCallbacks.h:163]: I don't know why this happened. I can only tell that a more up to date ZAP Tool seems to remove this lines.

File: src/controller/data_model/controller-clusters.zap (hunk 5 above)
Comment (eve-cxrp) [reply src/controller/data_model/controller-clusters.zap:739]: Sure I can do this. The only think is that I don't know which side effects this introduces. For example `src/app/zap-templates/zcl/data-model/silabs/ha-devices.xml` does reference the Binary Input Basic cluster.

**Review by vivien-apple — APPROVED** [review]
Approving for TE3 under the assumption that the changes mentioned in https://github.com/project-chip/connectedhomeip/pull/7032/files?file-filters%5B%5D=.h&file-filters%5B%5D=.js&file-filters%5B%5D=.m&file-filters%5B%5D=.mm&file-filters%5B%5D=.py#r638728211 are removed.

It is probably some conflicts into the ZAP file for something else since I did update ZAP into #7189 and have not noticed those changes.

It seems like the result of this PR is mostly to expose a set of accessors  to facilitate the manipulation of the `input basic` cluster attributes.
It makes perfect sense to facilitate it imho, but I would rather prefer to autogenerate helpers code such as what is proposed in #7183, instead of having a lot of new clusters introduced into `src/app/clusters`.

File: src/app/clusters/binary-input-server/binary-input-server.cpp [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+#include <support/logging/CHIPLogging.h>
+
+#ifndef emberAfBinaryInputBasicClusterPrintln
+#define emberAfBinaryInputBasicClusterPrintln(...) ChipLogProgress(Zcl, __VA_ARGS__);
+#endif
+
+EmberAfStatus emberAfBinaryInputBasicClusterGetPresentValue(chip::EndpointId endpoint, bool * presentValue)
+{
+    return emberAfReadServerAttribute(endpoint, ZCL_BINARY_INPUT_BASIC_CLUSTER_ID, ZCL_PRESENT_VALUE_ATTRIBUTE_ID,
+                                      (uint8_t *) presentValue, sizeof(uint8_t));
```
Comment (Damian-Nordic) [inline src/app/clusters/binary-input-server/binary-input-server.cpp:37]: `sizeof(*presentVaue)`?

**Review by Damian-Nordic — APPROVED** [review]

File: src/app/clusters/binary-input-server/binary-input-server.cpp (hunk 7 above)
Comment (eve-cxrp) [reply src/app/clusters/binary-input-server/binary-input-server.cpp:37]: I'm not sure this is okay to do. Because in C++ bool is not defined as uint8_t. Most compilers might implemented it in this manner but the saver way is to use uint8_t in my opinion.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/controller/data_model/gen/CHIPClientCallbacks.h (hunk 6 above)
Comment (bzbarsky-apple) [reply src/controller/data_model/gen/CHIPClientCallbacks.h]: It happened because of the pieces removed from  src/controller/data_model/controller-clusters.zap

File: src/app/zap-templates/templates/app/helper.js [hunk 8]
```diff
@@ -219,7 +221,7 @@ function chip_endpoint_generated_functions()
   let alreadySetCluster = [];
   let ret               = '\\\n';
   this.clusterList.forEach((c) => {
-    let clusterName  = extract_cluster_name(c.comment);
+    let clusterName  = c.clusterName;
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/helper.js:224]: Ah, this is dependent on the ZAP update in #7189.

**Review by bzbarsky-apple — COMMENTED** [review]
Also, codegen never got run across the tree here....

File: src/app/zap-templates/templates/app/helper.js [hunk 9]
```diff
@@ -195,8 +195,10 @@ function asChipUnderlyingType(label, type)
 // These helpers only works within the endpoint_config iterator
 
 // List of all cluster with generated functions
-var endpointClusterWithInit =
-    [ 'Basic', 'Identify', 'Groups', 'Scenes', 'Occupancy Sensing', 'On/off', 'Level Control', 'Color Control', 'IAS Zone' ];
+var endpointClusterWithInit = [
+  'Basic', 'Binary Input (Basic)', 'Identify', 'Groups', 'Scenes', 'Occupancy Sensing', 'On/off', 'Level Control', 'Color Control',
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/helper.js]: Fixed in commits I just pushed.

File: src/controller/data_model/controller-clusters.zap [hunk 10]
```diff
@@ -6859,36 +6956,6 @@
               "maxInterval": 65344,
               "reportableChange": 0
             },
-            {
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/controller/data_model/controller-clusters.zap]: Restored in commits I pushed.

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 11]
```diff
@@ -735,6 +735,103 @@
             }
           ]
         },
+        {
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.zap]: Moved to endpoint 1.

File: examples/all-clusters-app/all-clusters-common/all-clusters-app.zap [hunk 12]
```diff
@@ -735,6 +735,103 @@
...
+          "name": "Binary Input (Basic)",
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply examples/all-clusters-app/all-clusters-common/all-clusters-app.zap]: Rebased PR on ToT.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@eve-cxrp Can you take a look at the merge conflicts please?

@andy31415 @saurabhst @jelderton @Damian-Nordic ?

**Review by msandstedt — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
